src/mpid/common/self/mpidu_self.c

```diff
@@ -0,0 +1,308 @@
+            sdata_sz *= sreq->self_req.count;
+            MPIR_STATUS_SET_COUNT(*status, sdata_sz);
+            status->MPI_SOURCE = 0;
+            status->MPI_TAG = sreq->self_req.tag;
+            status->MPI_ERROR = MPI_SUCCESS;
+        }
+        *flag = TRUE;
+
+        *message = MPIR_Request_create(MPIR_REQUEST_KIND__MPROBE);
+        /* NOTE: we are repurposing the req field here. A union makes more sense, but hate
+         * the syntax clutter */
+        (*message)->self_req.req = sreq;
+        (*message)->comm = sreq->comm;  /* set so we can check it in MPI_{M,Im}recv */
+    } else {
+        *flag = FALSE;
+    }
+    MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_SELF_MUTEX);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_SELF_IMPROBE);
+    return MPI_SUCCESS;
+}
+
+int MPIDU_Self_imrecv(char *buf, MPI_Aint count, MPI_Datatype datatype,
+                      MPIR_Request * message, MPIR_Request ** request)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_SELF_IMRECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_SELF_IMRECV);
+    MPID_THREAD_CS_ENTER(VCI, MPIDIU_THREAD_SELF_MUTEX);
+
+    message->comm = NULL;       /* was set in MPIDU_Self_improbe */
+
+    MPIR_Request *sreq = message->self_req.req;
+    MPIR_Request *rreq = message;
+    rreq->kind = MPIR_REQUEST_KIND__RECV;
+    SELF_COPY(sreq->self_req.buf, sreq->self_req.count, sreq->self_req.datatype,
+              buf, count, datatype, rreq->status, sreq->self_req.tag);
+    MPIR_cc_set(&sreq->cc, 0);
+    /* comm will be released by MPIR_Request_free(sreq) */
+    MPIR_Datatype_release_if_not_builtin(sreq->self_req.datatype);
+    MPIR_cc_set(&rreq->cc, 0);
+
+    *request = rreq;
+    MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_SELF_MUTEX);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_SELF_IMRECV);
+    return MPI_SUCCESS;
+}
+
+#define DELETE_SELF(req_, found, queue) \
+    do { \
+        struct self_queue_entry *curr, *tmp; \
+        DL_FOREACH_SAFE(queue, curr, tmp) { \
+            if (curr->req == req_) { \
+                found = 1; \
+                DL_DELETE(queue, curr); \
+                break; \
+            } \
+        } \
+    } while (0)
+
+int MPIDU_Self_cancel(MPIR_Request * request)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_SELF_CANCEL);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_SELF_CANCEL);
+    MPID_THREAD_CS_ENTER(VCI, MPIDIU_THREAD_SELF_MUTEX);
+
+    if (!MPIR_Request_is_complete(request) && !MPIR_STATUS_GET_CANCEL_BIT(request->status)) {
+        int found = 0;
+        if (request->kind == MPIR_REQUEST_KIND__SEND) {
+            DELETE_SELF(request, found, self_send_queue);
+        } else {
+            DELETE_SELF(request, found, self_recv_queue);
+        }
+        if (found) {
+            /* comm will be released at MPIR_Request_free */
+            MPIR_Datatype_release_if_not_builtin(request->self_req.datatype);
+        } else {
+            /* error? */
```

Comment:
No error. A cancel will either succeed or fail and the user has to deal with it. Since this uses software matching, there's no problem with "un-cancelable" messages either.

Reply:
I'll just `goto fn_exit;` then.